### PR TITLE
Remove the zst allocation

### PR DESCRIPTION
**Pull request by oli-obk**

### CODE REVIEW

**Comment by RalfJung** [comment]
Heh, so you beat me to it. ;)

File: src/eval_context.rs [hunk 1]
```diff
@@ -362,15 +362,19 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
             StackPopCleanup::None => {},
             StackPopCleanup::Tls(key) => {
                 // either fetch the next dtor or start new from the beginning, if any are left with a non-null data
-                if let Some((instance, ptr, key)) = self.memory.fetch_tls_dtor(key).or_else(|| self.memory.fetch_tls_dtor(None)) {
+                let dtor = match self.memory.fetch_tls_dtor(key)? {
+                    dtor @ Some(_) => dtor,
+                    None => self.memory.fetch_tls_dtor(None)?,
+                };
+                if let Some((instance, ptr, key)) = dtor {
```
Comment (RalfJung) [inline src/eval_context.rs]: I have to say I find `or_else` clearer than an explicit match. Oh well.

File: tests/compile-fail/zst3.rs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// error-pattern: the evaluated program panicked
+
+#[derive(Debug)]
+struct A;
+
+fn main() {
+    // can't use assert_eq, b/c that will try to print the pointer addresses with full MIR enabled
```
Comment (RalfJung) [inline tests/compile-fail/zst3.rs]: Why didn't you use `assert!(... == ...)` (or `!=`) in the run-pass test?

**Comment by RalfJung** [comment]
So when a ZST local gets a heap allocation, we actually create a new empty allocation for it? Hm. Maybe that's the best we can do.

**Comment by eddyb** [comment]
ZST locals don't have "heap" allocations, they're not unlike C "local.objects" (vaguely "stack allocations").

**Comment by RalfJung** [comment]
Well, miri doesn't have an addressable stack, so -- the way I see it, it puts addressable local variables on the heap.

**Comment by eddyb** [comment]
it's only the host heap. The same allocation system is shared by the emulated heap, locals and globals.

**Comment by RalfJung** [comment]
*shrug* whatever you want to call it.

File: src/eval_context.rs (hunk 1 above)
Comment (oli-obk) [inline src/eval_context.rs]: I couldn't figure out how to do the early return

File: tests/compile-fail/zst3.rs (hunk 2 above)
Comment (oli-obk) [inline tests/compile-fail/zst3.rs]: Brainfart on my side...

File: src/eval_context.rs (hunk 1 above)
Comment (RalfJung) [inline src/eval_context.rs]: An, that requires some `Result::or_else` or so. Whatever.

File: tests/compile-fail/zst3.rs (hunk 2 above)
Comment (RalfJung) [inline tests/compile-fail/zst3.rs]: So you'll make a PR / push a fix for this?

Comment (oli-obk) [inline tests/compile-fail/zst3.rs]: Sure. Didn't think it was important

**Pull request merged**